### ISSUE TITLE
Provide more optimizers like Adam, RMSProp,AdaDelta, Adagrad, etc. ?

### ISSUE THREAD

**Issue by wangg12** [dmlc/mxnet#619]
As mentioned in #540, currently mxnet only has the normal SGD. It will be better to provide more excellent optimizing methods.

**Comment by antinucleon** [comment]
Thanks for suggestion! Indeed it is very easy to implement new optimization algorithm in MXNet eg, Adam in [Julia](https://github.com/dmlc/MXNet.jl/blob/master/src/optimizers/adam.jl). We may start to word on it after ICLR deadline. Also any contribution to make new optimizers to ```python/mxnet/optimizer.py``` is more than welcome :)

**Comment by winstywang** [comment]
Hi, thanks for the suggestion. Actually we have experimented with several advanced optimization methods in previous project cxxnet, however the difference between sgd is minor at least in the case of CNN. Anyway, we could implement some of them as mentioned by @antinucleon

**Comment by zachmayer** [comment]
I would also like to try out some different optimizers in the R packaged =D

**Comment by tqchen** [comment]
You are also more than welcomed to implement your own and contribute back to mxnet.

**Comment by zachmayer** [comment]
If you could provide some docs on writing new optimizers, I'll try my hand at porting adagrad from keras:
https://github.com/fchollet/keras/blob/a19c9ecfbd99a0fc60c476fb8a0fbd3284a9d620/keras/optimizers.py#L116

**Comment by thirdwing** [comment]
@zachmayer 

See https://github.com/dmlc/mxnet/blob/master/R-package/R/optimizer.R#L4 as an example.

**Comment by futurely** [comment]
Leslie N. Smith. No More Pesky Learning Rate Guessing Games. http://arxiv.org/abs/1506.01186.

**Comment by vchuravy** [comment]
@futurely That would be very nice to have.

**Comment by futurely** [comment]
Ilya Loshchilov, Frank Hutter. Online Batch Selection for Faster Training of Neural Networks. ICLR 2016. http://arxiv.org/abs/1511.06343
"Our experimental results on the MNIST dataset suggest that selecting batches speeds up both AdaDelta and Adam by a factor of about 5."

Code is available at https://sites.google.com/site/batchsel/.

**Comment by recastrodiaz** [comment]
I'm new to this framework and I'm wondering if there is any particular reason why optimizers are written in python/R/Julia? Other than that thanks for the great work!

I'd like to think that having them in C++ would improve code reuse, as binding languages wouldn't have to reimplement the optimizers again and again, just provide access to the C++ implementation. This could also improve performance (Try MNIST with 'sgd' and 'ccsgd').
The downside I see for having all optimizers in C++, would be that these would be harder to write, debug and maintain. (see https://github.com/dmlc/mxnet/blob/master/src/optimizer/sgd-inl.h vs https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L129).

Also, if anyone is interested in a C++ implementation of RMSProp 'ccrmsprop'. I can create a PR from my branch: https://github.com/dmlc/mxnet/compare/master...recastrodiaz:ccrmsprop

**Comment by zachmayer** [comment]
:+1: For writing a base set of optimizers in C++.  It'd be really nice to not have to figure out how to re-implement all the cool python stuff in R...

**Comment by piiswrong** [comment]
SGD optimizer is already in C++ now. Use with name 'ccsgd'. RMSProp is in python.

**Comment by futurely** [comment]
As @antinucleon [said](https://github.com/dmlc/mxnet/issues/1322#issuecomment-173435947), the Python version of RMSProp was broken. @recastrodiaz , would you like to PR your C++ version?

**Comment by rzoom** [comment]
@recastrodiaz I'd be very interested in trying out the ccrmsprop if you can get it merged.  MXNet is one of the few deep learning frameworks that is reasonable in C++, and having more choices for optimizers would be a great step forward.
